src/Common/src/System/Globalization/IdnMapping.Unix.cs

```diff
@@ -11,44 +11,39 @@ private unsafe string GetAsciiCore(char* unicode, int count)
             uint flags = Flags;
             CheckInvalidIdnCharacters(unicode, count, flags, nameof(unicode));
 
-            const int StackAllocThreshold = 512;
-            if (count < StackAllocThreshold)
-            {
-                char* output = stackalloc char[count];
-                return GetAsciiCore(unicode, count, flags, output, count, reattempt: true);
-            }
-            else
+            // Each unicode character is represented by 2 ASCII chars
+            // and a "xn--" prefix of length 4
+            const int StackallocThreshold = 512;
+            int estimatedLength = (int)Math.Min(count * 2L + 4, StackallocThreshold);
+            int actualLength;
+            if (estimatedLength < StackallocThreshold)
             {
-                char[] output = new char[count];
-                fixed (char* pOutput = output)
+                char* outputStack = stackalloc char[estimatedLength];
+                actualLength = Interop.GlobalizationNative.ToAscii(flags, unicode, count, outputStack, estimatedLength);
+                if (actualLength > 0 && actualLength <= estimatedLength)
```

Comment:
it is expected for non-ascii strings. I think we can mitigate this by increasing the allocations instead of (2 * count)+4 we can make it (3 * count) + 4. I believe this should still be reasonable as it'll will work with the short strings which is the main stream cases.